### PR TITLE
matrix/mat64,stat: reverse matrix extraction

### PR DESCRIPTION
**Pull request by kortschak**
Port of gonum/matrix#436.

@btracey Please take a look.

### CODE REVIEW

File: matrix/mat64/lq.go [hunk 1]
```diff
@@ -58,10 +58,15 @@ func (lq *LQ) Factorize(a Matrix) {
 // TODO(btracey): Add in the "Reduced" forms for extracting the m×m orthogonal
 // and upper triangular matrices.
 
-// LFromLQ extracts the m×n lower trapezoidal matrix from a LQ decomposition.
-func (m *Dense) LFromLQ(lq *LQ) {
+// LTo extracts the m×n lower trapezoidal matrix from a LQ decomposition.
```
Comment (btracey) [inline matrix/mat64/lq.go]: Do we need the `To` on the function calls? It's pretty clear from the function signature alone that it produces the matrix `L`, and very clear with the documentation. Seeing `x := lq.L(nil)` or `lq.L(x)` both feel well in line with our other signatures.

Comment (kortschak) [inline matrix/mat64/lq.go]: Thanks for reminding me. Yes, we do - see comments https://github.com/gonum/matrix/pull/436#issuecomment-287593489.

I will go through the other instances that are like this and change them to match. For example https://github.com/gonum/matrix/pull/433#discussion_r107590374.

Comment (btracey) [inline matrix/mat64/lq.go]: Thanks for reminding me. I still buy the argument.

**Comment by kortschak** [comment]
Added GSVD and HOGSVD equivalents. PTAL

(also some missed doc errors in previous commits)

File: matrix/mat64/gsvd.go [hunk 2]
```diff
@@ -214,112 +215,147 @@ func (m *Dense) ZeroRFromGSVD(gsvd *GSVD) {
...
-		m.set(i, i+k, 1)
+		dst.set(i, i+k, 1)
 	}
+	return dst
 }
 
-// UFromGSVD extracts the matrix U from the singular value decomposition, storing
-// the result in-place into the receiver. U is size r×r.
-func (m *Dense) UFromGSVD(gsvd *GSVD) {
+// UTo extracts the matrix U from the singular value decomposition, storing
```
Comment (kortschak) [inline matrix/mat64/gsvd.go]: > The documentation says U is rxr, but this makes it rxc.

U is (A_r,A_c), so U_r = U_c = r.

> The HOGSVD types have a "VTo will panic if the receiver does not contain a successful factorization." Should the GSVD types as well?

Will do.

File: matrix/mat64/gsvd.go [hunk 3]
```diff
@@ -214,112 +215,147 @@ func (m *Dense) ZeroRFromGSVD(gsvd *GSVD) {
...
 	}
-	m.Copy(tmp)
+	dst.Copy(tmp)
+	return dst
 }
 
-// VFromGSVD extracts the matrix V from the singular value decomposition, storing
-// the result in-place into the receiver. V is size p×p.
-func (m *Dense) VFromGSVD(gsvd *GSVD) {
+// VTo extracts the matrix V from the singular value decomposition, storing
```
Comment (kortschak) [inline matrix/mat64/gsvd.go]: See above.

File: matrix/mat64/gsvd.go [hunk 4]
```diff
@@ -214,112 +215,147 @@ func (m *Dense) ZeroRFromGSVD(gsvd *GSVD) {
...
 	}
-	m.Copy(tmp)
+	dst.Copy(tmp)
+	return dst
 }
 
-// QFromGSVD extracts the matrix Q from the singular value decomposition, storing
-// the result in-place into the receiver. Q is size c×c.
-func (m *Dense) QFromGSVD(gsvd *GSVD) {
+// QTo extracts the matrix Q from the singular value decomposition, storing
```
Comment (kortschak) [inline matrix/mat64/gsvd.go]: See above.

**Pull request merged**